[WIP] Settings feedback button changes

armaanahluwalia opened the pull request:
This PR is for changes to the save feedback styling and appearance on the settings overlay panel. It also refactors the code a little bit to ensure consistency and code cleanliness. I've made sure the changes have as minimal of a code footprint as possible.

*Tweaked look for the User Settings tabs*
![User Settings Feedback](http://g.recordit.co/85mgKcGxpC.gif)

*Tweaked Look for Organization Settings tabs*
![Organization Settings](http://g.recordit.co/TsKR8NGT7c.gif)

timabbott commented on static/js/settings_ui.js:
```diff
@@ -46,8 +49,10 @@ exports.do_settings_change = function (request_method, url, data, status_element
-            settings_ui.display_checkmark(spinner);
+            setTimeout(function() {
+              ui_report.success(success_msg, spinner, null, removeAfter);
+              settings_ui.display_checkmark(spinner);
```
We do 4-space indentation

timabbott commented on static/js/settings_ui.js:
```diff
@@ -33,10 +33,13 @@ exports.do_settings_change = function (request_method, url, data, status_element
         success_msg = opts.success_msg;
         success_continuation = opts.success_continuation;
         error_continuation = opts.error_continuation;
+        if(opts.sticky) removeAfter = null;
```
Stylistically, we generally do space after `if` and curly braces for all conditionals, for readability

timabbott commented on static/styles/settings.css:
```diff
@@ -351,22 +351,76 @@ input[type=checkbox] + .inline-block {
+}
+#settings_page .save-button {
+  margin-right: 5px;
+}
```
Stylistically, we do blank lines between blocks in our CSS

timabbott commented on static/templates/settings/settings-save-discard-widget.handlebars:
```diff
@@ -1,12 +1,22 @@
+        {{t 'Discard' }}
+      </span>
+    </div>
+  </div>
```
I think the linter probably complained about this (`tools/lint`) but we generally do 4-space indentation in HTML templates in this codebase.

timabbott commented on static/js/settings_org.js:
```diff
@@ -583,10 +608,10 @@ function _set_up() {
         e.preventDefault();
         e.stopPropagation();
-        var save_button = $(e.target);
+        var save_button = $(e.currentTarget);
```
What's the story with this change?

shubhamdhama replied:
I guess with `currentTarget` we are sure that the save button will refer to `.subsection-header .subsection-changes-save .button`, but here I can't find any solid reason to prefer this over `target`.

armaanahluwalia replied:
currentTarget is used because with the changes the .button container element has both a icon span and a text span. Using target would refer to these elements instead of the .button container element.

timabbott replied:
OK, makes sense.

timabbott commented on static/js/settings_org.js:
```diff
@@ -450,6 +450,39 @@ function _set_up() {
+      } else if(state === 'succeeded') {
+        showHideElement('hide');
+      }
+    }
```
I think you could probably split this commit into 2: first a commit that extracts this function (without changing behavior; just moving the code around), and then a second that changes things.  That might be more readable.

armaanahluwalia replied:
Hi tim - I'm not sure if I completely understand what you mean in your comment. This is a new function written to put all save-discard button state changes into one function that could ( and likely should at some point be extracted out into its own file ). When you say move code around I don't completely follow.

timabbott commented:
Also CC @shubhamdhama who can probably also provide a useful review in the India timezone.

shubhamdhama commented:
This looks great, here are my few thoughts about the changes:
- In org settings, **loading indicator** looks nice in button. But visually it looks odd as compared to other buttons. Here are my suggestions, 
  - Can we implement this indicator in previous buttons? (Actually, I've also tried them with buttons and there were some major issues with CSS but it seems it is implemented in a better way here).
  - (feedback required) Another thing we can do here is to hide the buttons on clicking save and show the same "saving status indicator" which we have implemented in display and notifications settings.
- There are some minor issues with **fading** out of saving indicator when checkboxes are checked quickly and randomly(which I don't think a user will do usually). Also, there was a related issue/PR regarding "hiding of status indicator after an interval" which was closed so I think there can be some problems with this approach.
- Mostly, we prefer `lowercase_with_underscore` over `camelCase`.
- I'm positive about font-awesome loading indicator, not clear why we haven't used it before.

For how it looks I think it's better to concern @rishig.

armaanahluwalia commented:
Thanks for the comments @shubhamdhama. Will wait to hear what consensus is reached and also to hear from @rishig.

rishig commented:
Thanks @armaanahluwalia! Some visual design notes:

User Settings
* The checkmark seems a bit off vertically? I feel like I might also add another pixel of space between the loading circle and the text.

Organization settings:
These buttons are quite different from the style used everywhere else in the product. How about we make the discard button similar to our current black button style (maybe with an X icon added in), and then change the "Save changes" button appropriately? (E.g. sentence casing, lower font weight, etc., but keep the solid background.)

armaanahluwalia commented:
Thanks for commenting @rishig . Could you point me to the black button style you're referring to?

armaanahluwalia commented:
Hi @rishig @timabbott @shubhamdhama I've updated this PR based on your comments. 

Here is how it looks now. I hope it better reflects the current Zulip style.
![User Settings Feedback](http://g.recordit.co/mgHxiDYKNA.gif)
![Organization Settings Buttons](http://g.recordit.co/2S7BsGI3SW.gif)

My apologies for not running it through linter earlier I missed that. I've updated all those issues.
I also fixed a bug thats currently in production where if you click on a setting that requires reload, and then click on a setting that doesn't require reload, the reload message is overwritten by the plain saved message. This could be confusing to users since they won't be aware that they require reload to see changes. I've written some code that corrects that.

@shubhamdhama Apart from the above bug I really did not find any issues with the fade and quickly selecting different options. Please let me know if there are any issues you see.

One thing I intentionally did not incorporate is ** not camelCasing vars in JavaScript**
Shubham did mention that you prefer casing with underscores, however I looked through the JavaScript Coding standards and I see that's not really in there. I was very relieved to see that since camelCasing variables in JavaScript is really the industry standard and theres not a lot of debate around this. I would be happy if this was at least up to me to decide, however I am also happy to go in and rebase the code again to incorporate this change if there strong sentiment around it.

rishig commented:
The design looks great! As followup we may want to decide on a consistent checkmark style (and maybe have a larger conversation about button styles), but that doesn't need to block this PR.

For camelCasing: we can start a separate discussion about migrating our js codebase to camelCase, but we shouldn't combine that discussion with this PR. So for now I would go with whatever is consistent with the rest of the codebase, and we can have the larger casing discussion in #frontend.

armaanahluwalia commented:
Hi @rishig @shubhamdhama  I've updated this PR with the 2 points you mentioned. I've edited all the new code to now include camelCase.

I've also gone ahead and implemented a better and more consistent checkmark style by introducing a new icon from Google's material design kit. See attached GIFs. Personally, I feel its more consistent between settings and organization and looks comparably good to the old one.

If you don't like this new style I can drop the last commit to include only casing changes. I believe that should conclude the code changes needed for this PR unless you have any further thoughts.

![Checkmark on User Settings](http://g.recordit.co/3TmoBvoSql.gif)
![Buttons on Organization Settings](http://g.recordit.co/qUBTXjfDw2.gif)

rishig commented:
cool! I think a steeper checkmark here would look nicer, e.g. similar to the checkmark used in the checkboxes? Also possibly the checkmark and `x` are a bit low in the second gif, e.g. relative to the first gif.

armaanahluwalia commented:
Hi @rishig . I've changed the checkmark to be steeper as requested. I hope this looks better to you. If not, I will roll back to the previous commit and let these design changes be brought in a different PR as you had suggested earlier. Please see attached GIFs. 

![User Settings Save Button](http://g.recordit.co/hMhYandrXg.gif)
![Organization Settings Button](http://g.recordit.co/OjP51Q2IsL.gif)

I've also added tests for the newly added code so I hope this PR is ready for merge now : )

timabbott commented on static/styles/settings.css:
```diff
@@ -471,14 +552,13 @@ input[type=checkbox].inline-block {
-
     margin-top: 14px;
     margin-left: 10px;
+    border: 1px solid #59a687;
```
the reason we've been using e.g. `hsl(178, 100%, 40%);` to declare colors is so that they work sensibly in "night mode" without needing to add styles in `night_mode.css` for everything.  Try toggling that setting and you'll see what I mean.

timabbott commented on static/js/settings_org.js:
```diff
@@ -422,46 +462,38 @@ function _set_up() {
-                save_button.removeClass('hide').addClass('show').stop(true).fadeTo(0, 1);
+                setTimeout(function () {
+                  exports.change_save_button_state(save_btn_container, "succeeded");
+                }, 500);
```
Is there a reason you didn't put the `setTimeout` inside `export.change_save_button_state`?

armaanahluwalia replied:
I mostly kept the setTimeout outside the change_save_button_state for testability and keeping the function tight. This way it can be used synchronously should the need arise. I imagine helper functions like this being shared across different sections/modules once things become a little more consistent in combination with the associated template.

timabbott commented:
I posted a few quick comments.  Can you rebase (should be able to just `git rebase --skip` the 3 commits I merged), and fix those comments?  I'm happy to merge at that point, and we can fix-forward anything else from there.

armaanahluwalia commented:
@timabbott I rebased and changed the colors to HSL.

timabbott commented:
I fixed a few more code style things that our linter should probably enforce and merged, thanks for all your work on this @armaanahluwalia!  

@shubhamdhama FYI; you may want to play around with this looking for follow-up things we might want to fix, since this is sorta your subsystem :).  The main thing I noticed is that I think some of the colors are a bit inconsistent in night mode.
